p11-kit/rpc-message.c

```diff
@@ -1660,6 +1668,386 @@ p11_rpc_buffer_get_ibm_attrbound_wrap_mechanism_value (p11_buffer *buffer,
+{
+	const unsigned char *data;
+	unsigned char has_data;
+	size_t len;
+	uint32_t length;
+
+	if (!p11_rpc_buffer_get_byte(buffer, offset, &has_data))
+		return false;
+
+	if (has_data == 0) {
+		if (!p11_rpc_buffer_get_uint32(buffer, offset, &length))
+			return false;
+		len = length;
+	} else {
+		if (!p11_rpc_buffer_get_byte_array(buffer, offset, &data, &len))
+			return false;
+	}
+
+	if (value) {
+		CK_IBM_BTC_DERIVE_PARAMS *params = (CK_IBM_BTC_DERIVE_PARAMS *) value;
+
+		if (params->pChainCode && params->ulChainCodeLen == len) {
+			memcpy(params->pChainCode, data, len);
+			params->ulChainCodeLen = len;
+		} else {
+			params->pChainCode = (void *) data;
+			params->ulChainCodeLen = len;
+		}
+	}
+
+	if (value_length)
+		*value_length = sizeof (CK_IBM_BTC_DERIVE_PARAMS);
+
+	return true;
+}
+
+void
+p11_rpc_buffer_add_ibm_ecdsa_other_mechanism_value (p11_buffer *buffer,
+						    const void *value,
+						    CK_ULONG value_length)
+{
+	CK_IBM_ECDSA_OTHER_PARAMS params;
+
+	if (value_length != sizeof (CK_IBM_ECDSA_OTHER_PARAMS)) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+
+	memcpy (&params, value, value_length);
+
+	if (params.submechanism > UINT64_MAX) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+
+	p11_rpc_buffer_add_uint64 (buffer, params.submechanism);
+}
+
+bool
+p11_rpc_buffer_get_ibm_ecdsa_other_mechanism_value (p11_buffer *buffer,
+					    size_t *offset,
+					    void *value,
+					    CK_ULONG *value_length)
+{
+	uint64_t val1;
+
+	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val1))
+		return false;
+
+	if (value) {
+		CK_IBM_ECDSA_OTHER_PARAMS params;
+
+		params.submechanism = val1;
+
+		memcpy (value, &params, sizeof (CK_IBM_ECDSA_OTHER_PARAMS));
+	}
+
+	if (value_length)
+		*value_length = sizeof (CK_IBM_ECDSA_OTHER_PARAMS);
+
+	return true;
+}
+
+void
+p11_rpc_buffer_add_ibm_btc_derive_mechanism_value (p11_buffer *buffer,
+						       const void *value,
+						       CK_ULONG value_length)
+{
+	CK_IBM_BTC_DERIVE_PARAMS params;
+
+	if (value_length != sizeof (CK_IBM_BTC_DERIVE_PARAMS)) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+
+	memcpy (&params, value, value_length);
+
+	if (params.type > UINT64_MAX) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+	p11_rpc_buffer_add_uint64(buffer, params.type);
+
+	if (params.childKeyIndex > UINT64_MAX) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+	p11_rpc_buffer_add_uint64(buffer, params.childKeyIndex);
+
+	p11_rpc_buffer_add_byte_array(buffer, (unsigned char *)params.pChainCode, params.ulChainCodeLen);
+
+	if (params.version > UINT64_MAX) {
+		p11_buffer_fail (buffer);
+		return;
+	}
+	p11_rpc_buffer_add_uint64(buffer, params.version);
+}
+
+bool
+p11_rpc_buffer_get_ibm_btc_derive_mechanism_value (p11_buffer *buffer,
+						   size_t *offset,
+						   void *value,
+						   CK_ULONG *value_length)
+{
+	uint64_t val1;
+	uint64_t val2;
+	size_t len;
+	const unsigned char *data;
+	uint64_t val3;
+
+	if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val1))
```

Comment:
You can combine these into a single `if` which would look much nicer, something like:
```
if (!p11_rpc_buffer_get_uint64 (buffer, offset, &val1) ||
    !p11_rpc_buffer_get_uint64 (buffer, offset, &val2) ||
     ...)
      return false;
```
Same can be done in `p11_rpc_buffer_get_ibm_kyber_mechanism_value`